### ISSUE TITLE
Psalm may fail to warn if passed invalid key type for array access on an array

### ISSUE THREAD

**Issue by TysonAndre** [vimeo/psalm#309]
This is a feature that I'd want psalm to have, because it has array key types as a part of union types.

When `int`, `int|string`, `string`, or `mixed` are the array key types, I'd expect more warnings for code such as the below if the key can't cast to int|string:

```php
<?php
/**
 * @property array<int,int> $prop
 */
class A{
    /** @param string $x */
    public function __get($x) {
        return [2, $x];
    }
}
// This is incorrect code, designed to test Psalm's type checker on array keys
$c = new A();
echo strlen($c->prop);  // correctly warns $prop is array<int, int>
echo $c->prop['key'];

// Also, psalm isn't warning about array keys for regular arrays:
$x = ['value'];
var_export($x['key']);
var_export($x[[]]);  // still doesn't warn
var_export(strlen($x));  // array<int, string> provided
```

**Comment by TysonAndre** [comment]
What are $inferred_key_type and $key_type supposed to be in src/Psalm/Checker/Statements/Expression/FetchChecker.php (analyzeArrayAccess)? I see that $key_type gets modified to include both types from the array and type from the key's expression.

It seems like $key_type is first created as the inferred type of $stmt->dim, then reused for the types expected by the array/array-like $stmt->var, which seems wrong.

After ExpressionChecker::analyze(), could a new union type be created for the expected types for $stmt->var?

**Comment by muglug** [comment]
Lots and lots of bugs found having fixed this locally. Thanks!

**Comment by TysonAndre** [comment]
I'm still having issues with magic properties, e.g. echo $c->prop['key']; should expect an int for an array key.

**Comment by muglug** [comment]
> I'm still having issues with magic properties, e.g. echo $c->prop['key']; should expect an int for an array key.

Still having those issues?

**Comment by muglug** [comment]
This was a pandora's box of a bug – once fixed, a WHOLE bunch of bugs related to the handling of object-like arrays bubbled up, and now Psalm is able to find a ton of bugs related to the heavy (over)use of arrays in our codebase. So thanks again for reporting.

**Comment by TysonAndre** [comment]
> > I'm still having issues with magic properties, e.g. echo $c->prop['key']; should expect an int for an array key.

> Still having those issues?

No, those are fixed. Other issues were caused by `@property-read` being ignored and `null`/`string` (I forget which) being one of the possible element types. I fixed that on my fork, and should be able to fix the PR tomorrow
